skrf/media/media.py

```diff
@@ -833,24 +838,74 @@ def line(self, d: NumberLike, unit: str = 'deg',
+                          'to determine if the line has to be renormalized.',
+              FutureWarning, stacklevel = 2)
             # Use the same s_def here as the line to avoid changing it during
             # cascade.
-            result = self.thru(s_def='traveling')**result**self.thru(s_def='traveling')
+            if not (z0 is None) and not (self.z0 is None):
+                kwargs.update({'z0' : z0})
+                s_def = kwargs.pop('s_def', S_DEF_DEFAULT)
+                result = self.match(nports=2, s_def='traveling', **kwargs)
+                theta = self.electrical_length(self.to_meters(d=d, unit=unit))
+                s11 = npy.zeros(self.frequency.npoints, dtype=complex)
+                s21 = npy.exp(-1*theta)
+                result.s = \
+                        npy.array([[s11, s21],[s21,s11]]).transpose().reshape(-1,2,2)
+                z_renormalize = fix_z0_shape(self.z0, self.npoints, 2)
+                result.renormalize(z_renormalize, s_def='traveling')
+            else:
+                raise ValueError('Missing z0 or Z0 impedance specification')
+        
+        else:
+            # priority given to line impedances on media impedances.
+            z_port = self.z0 if z0 is None else z0
+            z_char = self.Z0 if Z0 is None else Z0
+            
+            # z shape has to be fixed for renormalize to works properly
+            # or it will trigger
+            # AttributeError: 'bool' object has no attribute 'any'
+            # because of `if (self.z0 != z_new).any():` on line 3033
+            if z_port is None:
+                if z_char is None:
+                    raise ValueError('Neither port or characteristic impedance specified')
+                else:
+                    z_match = fix_z0_shape(z_char, self.npoints, 2)
+                    z_renormalize = None     
+            else:
+                if z_char is None:
+                    z_match = fix_z0_shape(z_port, self.npoints, 2)
+                    z_renormalize = None
+                else:
+                    z_match = fix_z0_shape(z_char, self.npoints, 2)
+                    z_renormalize = fix_z0_shape(z_port, self.npoints, 2)
+                
+            # match
+            kwargs.update({'z0' : z_match})
+            s_def = kwargs.pop('s_def', S_DEF_DEFAULT)
+            # Need to use either traveling or pseudo definition here
+            # for the network to match traveling waves.
+            result = self.match(nports=2, s_def='traveling', **kwargs)
+    
+            # transmission line
+            theta = self.electrical_length(self.to_meters(d=d, unit=unit))
+            s11 = npy.zeros(self.frequency.npoints, dtype=complex)
+            s21 = npy.exp(-1*theta)
+            result.s = \
+                npy.array([[s11, s21],[s21,s11]]).transpose().reshape(-1,2,2)
+            
+            if not (z_renormalize is None):
+                result.renormalize(z_renormalize, s_def='traveling')
```

Comment:
There is another renormalize just after this one for s_def change. It can be removed and `s_def` set here (and other if branch) to avoid renormalizing twice.